cmd/codegen/generator/go/generator.go

```diff
@@ -317,7 +317,9 @@ func loadPackage(ctx context.Context, dir string) (*packages.Package, *token.Fil
 			packages.NeedTypes |
 			packages.NeedSyntax |
 			packages.NeedTypesInfo |
-			packages.NeedModule,
+			packages.NeedModule |
+			packages.NeedDeps |
+			packages.NeedImports,
```

Comment:
I left these ones out in the first iteration of all this because there was a scary-sounding warning in this blog post about it potentially being quite expensive: https://eli.thegreenplace.net/2020/writing-multi-package-analysis-tools-for-go/ (search `This is an expensive operation, as you can imagine`)

Can we attempt a stress test of this by adding a ton of imports (preferably those with a lot of transitive deps) and ensures the we don't use some absurd amount of memory and/or take forever to load the module?

If it is a problem, there's probably ways of dealing with it, but just curious to what degree it even is an issue (if at all).

Reply:
> Sigh. Yeah, I worried a bit about adding this

I do think it's worth checking how much of an issue this is in practice if you haven't already. "an expensive operation" is a highly relative/subjective thing and also just a by-line in a blog post (though I do trust the author as pretty knowledgeable, so also not worth dismissing).

If particularly extreme cases only add a few megs of memory and/or a few seconds of time then I would probably be inclined to not especially care yet and we can just add this imo.

Reply:
Ok, I added a note inline about this maybe being expensive.

I tried a test for this:

```go
func TestModuleGoPackageStress(t *testing.T) {
        t.Parallel()

        c, ctx := connect(t)

        modGen := c.Container().From(golangImage).
                WithMountedFile(testCLIBinPath, daggerCliFile(t, c)).
                WithWorkdir("/work").
                With(daggerExec("mod", "init", "--name=foo", "--sdk=go")).
                WithNewFile("main.go", dagger.ContainerWithNewFileOpts{
                        Contents: `package main

import (
        _ "github.com/moby/buildkit/cache"
        _ "github.com/moby/buildkit/solver"
        _ "github.com/moby/buildkit/worker/containerd"
)

type Foo struct {}

func (m *Foo) Hello() string {
        return "Hello!"
}
`,
                })
        logGen(ctx, t, modGen.Directory("."))

        out, err := modGen.With(daggerQuery(`{foo{hello}}`)).Stdout(ctx)
        require.NoError(t, err)
        require.JSONEq(t, `{"foo": {"hello": "Hello!"}}`, out)
}
```

This pulls in quite a few deps (120 of them):

```
go: downloading github.com/containerd/containerd v1.7.2
go: downloading github.com/Microsoft/go-winio v0.6.1
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/grpc-ecosystem/go-grpc-middleware v1.3.0
go: downloading github.com/containerd/nydus-snapshotter v0.8.2
go: downloading github.com/containerd/typeurl/v2 v2.1.1
go: downloading github.com/docker/docker v24.0.0-rc.2.0.20230718135204-8e51b8b59cb8+incompatible
go: downloading github.com/sirupsen/logrus v1.9.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/opencontainers/image-spec v1.1.0-rc3
go: downloading github.com/golang/protobuf v1.5.3
go: downloading go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc v0.40.0
go: downloading go.opentelemetry.io/otel v1.14.0
go: downloading google.golang.org/grpc v1.53.0
go: downloading github.com/opencontainers/go-digest v1.0.0
go: downloading go.opentelemetry.io/otel/trace v1.14.0
go: downloading golang.org/x/net v0.8.0
go: downloading go.etcd.io/bbolt v1.3.7
go: downloading github.com/moby/sys/mountinfo v0.6.2
go: downloading github.com/containerd/continuity v0.4.1
go: downloading github.com/tonistiigi/fsutil v0.0.0-20230629203738-36ef4d8c0dbb
go: downloading github.com/containerd/stargz-snapshotter/estargz v0.14.3
go: downloading github.com/opencontainers/runtime-spec v1.1.0-rc.2
go: downloading go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.14.0
go: downloading github.com/klauspost/compress v1.17.2
go: downloading go.opentelemetry.io/otel/sdk v1.14.0
go: downloading github.com/mitchellh/hashstructure/v2 v2.0.2
go: downloading github.com/opencontainers/runc v1.1.7
go: downloading github.com/opencontainers/selinux v1.11.0
go: downloading github.com/docker/distribution v2.8.2+incompatible
go: downloading github.com/containerd/stargz-snapshotter v0.14.3
go: downloading github.com/in-toto/in-toto-golang v0.5.0
go: downloading golang.org/x/crypto v0.2.0
go: downloading github.com/spdx/tools-golang v0.5.1
go: downloading golang.org/x/sys v0.13.0
go: downloading golang.org/x/tools v0.14.0
go: downloading github.com/gogo/googleapis v1.4.1
go: downloading github.com/AdamKorcz/go-118-fuzz-build v0.0.0-20221215162035-5330a85ea652
go: downloading go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace v0.40.0
go: downloading go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp v0.40.0
go: downloading google.golang.org/genproto v0.0.0-20230306155012-7f2fa6fef1f4
go: downloading google.golang.org/protobuf v1.30.0
go: downloading github.com/containerd/fifo v1.1.0
go: downloading github.com/moby/sys/signal v0.7.0
go: downloading github.com/Microsoft/hcsshim v0.10.0-rc.8
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading kernel.org/pub/linux/libs/security/libcap/cap v1.2.67
go: downloading github.com/gofrs/flock v0.8.1
go: downloading github.com/moby/locker v1.0.1
go: downloading github.com/armon/circbuf v0.0.0-20190214190532-5111143e8da2
go: downloading github.com/tonistiigi/units v0.0.0-20180711220420-6950e57a87ea
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading go.opentelemetry.io/otel/metric v0.37.0
go: downloading go.opentelemetry.io/proto/otlp v0.19.0
go: downloading github.com/containerd/ttrpc v1.2.2
go: downloading github.com/google/go-cmp v0.5.9
go: downloading github.com/AdaLogics/go-fuzz-headers v0.0.0-20230106234847-43070de90fa1
go: downloading github.com/moby/patternmatcher v0.5.0
go: downloading github.com/Masterminds/semver/v3 v3.1.0
go: downloading github.com/go-logr/logr v1.2.3
go: downloading github.com/containerd/go-cni v1.1.9
go: downloading github.com/moby/sys/sequential v0.5.0
go: downloading github.com/prometheus/procfs v0.9.0
go: downloading github.com/package-url/packageurl-go v0.1.1-0.20220428063043-89078438f170
go: downloading github.com/vbatts/tar-split v0.11.2
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/secure-systems-lab/go-securesystemslib v0.4.0
go: downloading github.com/shibumi/go-pathspec v1.3.0
go: downloading github.com/felixge/httpsnoop v1.0.3
go: downloading golang.org/x/time v0.3.0
go: downloading github.com/docker/go-units v0.5.0
go: downloading github.com/pelletier/go-toml v1.9.5
go: downloading github.com/docker/go-connections v0.4.0
go: downloading github.com/tonistiigi/go-archvariant v1.0.0
go: downloading github.com/go-logr/stdr v1.2.2
go: downloading golang.org/x/mod v0.13.0
go: downloading golang.org/x/text v0.8.0
go: downloading gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c
go: downloading kernel.org/pub/linux/libs/security/libcap/psx v1.2.67
go: downloading github.com/docker/go-events v0.0.0-20190806004212-e31b211e4f1c
go: downloading github.com/grpc-ecosystem/grpc-gateway/v2 v2.11.3
go: downloading github.com/anchore/go-struct-converter v0.0.0-20221118182256-c68fdcfa2092
go: downloading github.com/cyphar/filepath-securejoin v0.2.3
go: downloading github.com/hashicorp/go-immutable-radix v1.3.1
go: downloading github.com/containernetworking/cni v1.1.2
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading go.opencensus.io v0.24.0
go: downloading github.com/kr/pretty v0.3.1
go: downloading github.com/moby/sys/mount v0.3.3
go: downloading github.com/stretchr/objx v0.5.0
go: downloading github.com/containerd/cgroups v1.1.0
go: downloading github.com/rogpeppe/go-internal v1.11.0
go: downloading github.com/kr/text v0.2.0
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading golang.org/x/net v0.16.0
go: downloading gotest.tools v2.2.0+incompatible
go: downloading golang.org/x/crypto v0.14.0
go: downloading golang.org/x/oauth2 v0.6.0
go: downloading github.com/codahale/rfc6979 v0.0.0-20141003034818-6a90f24967eb
go: downloading github.com/onsi/ginkgo/v2 v2.1.3
go: downloading github.com/onsi/gomega v1.17.0
go: downloading github.com/hashicorp/go-uuid v1.0.0
go: downloading github.com/golang/glog v1.0.0
go: downloading golang.org/x/term v0.13.0
go: downloading google.golang.org/appengine v1.6.7
go: downloading cloud.google.com/go/compute/metadata v0.2.3
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/onsi/ginkgo v1.16.4
go: downloading cloud.google.com/go/compute v1.18.0
go: downloading golang.org/x/text v0.13.0
go: downloading cloud.google.com/go v0.110.0
go: downloading gotest.tools/v3 v3.5.1
```

Running my stress test (which I haven't committed to the code since there's not really an easy way to track this over time) on main vs on this branch I get 34.96s on main and 35.692s on this branch - maybe there's a couple of seconds of difference on test runtime?

I'd be happy to pull this change in now, and revisit later if we start seeing performance issues in the future? I'm not convinced there's enough of a perf issue to spend dev cycles now - especially since this cost should only be paid on first run, on future cases this should be cached in the go cache directory.